README.md
wot
===

What's On TV

Ruby script to process Australian TV guide

Usage
=====

```
wot.rb -[h|C|L|n|N|r] -c Channel -a Time -l Location -s Search

-h: Display usage
-c: Channel [TEN,ABC1,..]
-a: Time [8,8.00,8:30,..]
-l: Location (Default Melbourne)
-C: List channels
-L: List locations
-n: What's on TV now (the current hour)
-N: What's on TV next (the next hour)
-R: Want's on TV (the rest of the day)
-s: Search on subject (eg News)
-r: Show only staff picks
```

Examples
========

Display Channel TEN's programming for the rest of the day:

```
$ wot.rb -c TEN
+----------------------------+-------------+-------------+-------------+
|        TV Programme: TEN Melbourne (Tuesday 22 October 2013)         |
+----------------------------+-------------+-------------+-------------+
| Program                    | Channel     | Time        | Staff Pick  |
+----------------------------+-------------+-------------+-------------+
| Dr. Phil                   | TEN         | 12pm        |             |
| The Doctors                | TEN         | 1pm         |             |
| Ready Steady Cook          | TEN         | 2pm         |             |
| Judge Judy                 | TEN         | 3pm         |             |
| Huey's Kitchen             | TEN         | 3.30pm      |             |
| Totally Wild               | TEN         | 4pm         |             |
| The Bold and the Beautiful | TEN         | 4.30pm      |             |
| TEN Eyewitness News        | TEN         | 5pm         |             |
| The Simpsons               | TEN         | 6pm         |             |
| The Project                | TEN         | 6.30pm      |             |
| Recipe to Riches           | TEN         | 7.30pm      |             |
| NCIS                       | TEN         | 8.30pm      | Yes         |
| NCIS: Los Angeles          | TEN         | 9.30pm      |             |
| TEN Eyewitness News Late   | TEN         | 10.30pm     |             |
| The Project                | TEN         | 11.15pm     |             |
+----------------------------+-------------+-------------+-------------+
```

Display what is on TV at 6:30pm:

```
$ wot.rb -a 6:30pm
+------------------------------+---------------+---------------+---------------+
|        TV Programme:  Melbourne at 6.30 pm (Tuesday 22 October 2013)         |
+------------------------------+---------------+---------------+---------------+
| Program                      | Channel       | Time          | Staff Pick    |
+------------------------------+---------------+---------------+---------------+
| World News Australia         | SBSONE        | 6.30pm        |               |
| World News Australia         | SBSHD         | 6.30pm        |               |
| Desperate Measures           | NITV          | 6.30pm        |               |
| Today Tonight                | Seven         | 6.30pm        |               |
| How I Met Your Mother        | 7mate         | 6.30pm        |               |
| A Current Affair             | Nine          | 6.30pm        |               |
| Friends                      | GEM           | 6.30pm        |               |
| The Middle                   | GO!           | 6.30pm        |               |
| Get Smart                    | ONE           | 6.30pm        |               |
| The Project                  | TEN           | 6.30pm        |               |
| Neighbours                   | ELEVEN        | 6.30pm        |               |
| Total Gym XLS                | SpreeTV       | 6.30pm        |               |
| PURMinerals Makeup GOES WILD | TVSN          | 6.30pm        |               |
| euromaxx                     | TVS           | 6.30pm        |               |
| No More Practice             | 4ME           | 6.30pm        |               |
+------------------------------+---------------+---------------+---------------+
```

Search for type of programming:

```
$ wot.rb -s "Movie"
+---------------+--------------+--------------+--------------+
|  TV Programme:  Melbourne Movie (Tuesday 22 October 2013)  |
+---------------+--------------+--------------+--------------+
| Program       | Channel      | Time         | Staff Pick   |
+---------------+--------------+--------------+--------------+
| At the Movies | ABC1         | 9.31pm       |              |
+---------------+--------------+--------------+--------------+
```

List channels:

```
$ wot.rb -C
Channels:
ABC1
ABC2 / ABC4
ABC3
ABC News 24
SBS ONE
SBS HD
SBS 2
NITV
...
```

List locations:

```
$ wot.rb -L
Locations:
Sydney
Melbourne
Brisbane
Adelaide
Darwin
Hobart
Perth
```

Show staff picks:

```
$ wot.rb -r
+------------------------+-------------+-------------+-------------+
|        TV Programme:  Melbourne (Tuesday 22 October 2013)        |
+------------------------+-------------+-------------+-------------+
| Program                | Channel     | Time        | Staff Pick  |
+------------------------+-------------+-------------+-------------+
| The Art of Australia   | ABC1        | 8.32pm      | Yes         |
| Lost Girl              | SBS2        | 8.40pm      | Yes         |
| Dancing with the Stars | Seven       | 7.30pm      | Yes         |
| Super Fun Night        | Nine        | 8pm         | Yes         |
| NCIS                   | TEN         | 8.30pm      | Yes         |
+------------------------+-------------+-------------+-------------+
```

Show what's on for the rest of the day:

```
$ wor.rb -R
+----------------------------------------+------------+------------+------------+
|                TV Programme:  Melbourne (Tuesday 20 May 2014)                 |
+----------------------------------------+------------+------------+------------+
| Program                                | Channel    | Time       | Staff Pick |
+----------------------------------------+------------+------------+------------+
| The Writers' Room                      | ABC1       | 10pm       |            |
| Lateline                               | ABC1       | 10.23pm    |            |
| The Business                           | ABC1       | 10.58pm    |            |
| Four Corners                           | ABC1       | 11.25pm    |            |
| Outside Chance                         | ABC2/ABC4  | 9.36pm     |            |
| Archer                                 | ABC2/ABC4  | 10.05pm    |            |
| Gruen Planet                           | ABC2/ABC4  | 10.28pm    |            |
| The Roast                              | ABC2/ABC4  | 11.06pm    |            |
| Ross Kemp: Extreme World               | ABC2/ABC4  | 11.18pm    |            |
| Station Close                          | ABC3       | 9.02pm     |            |
| The World                              | ABCNews24  | 10pm       |            |
| ABC News                               | ABCNews24  | 11pm       |            |
| 7.30                                   | ABCNews24  | 11.30pm    |            |
| Dateline                               | SBSONE     | 9.30pm     |            |
| SBS World News                         | SBSONE     | 10.30pm    |            |
| Giro d'Italia                          | SBSONE     | 11pm       |            |
| Dateline                               | SBSHD      | 9.30pm     |            |
| SBS World News                         | SBSHD      | 10.30pm    |            |
| Giro d'Italia                          | SBSHD      | 11pm       |            |
| The Walking Dead                       | SBS2       | 9.20pm     |            |
| The Feed                               | SBS2       | 10.10pm    |            |
| Russell Howard's Good News             | SBS2       | 10.40pm    |            |
| Ichi                                   | SBS2       | 11.15pm    |            |
| ANZ Netball Championship 2014          | NITV       | 10pm       |            |
| Coldplay: Ghost Stories                | Seven      | 9.45pm     |            |
| Ghost                                  | Seven      | 11pm       |            |
| Air Crash Investigation                | 7TWO       | 9.30pm     |            |
| Seconds From Disaster                  | 7TWO       | 10.40pm    |            |
| When Weather Changed History           | 7TWO       | 11.45pm    |            |
| Family Guy                             | 7mate      | 10pm       |            |
| Family Guy                             | 7mate      | 10.30pm    |            |
| American Dad!                          | 7mate      | 11pm       |            |
| American Dad!                          | 7mate      | 11.30pm    |            |
| Mike's Extreme Training                | 4Me        | 10pm       |            |
| Pilates in Your Home!                  | 4Me        | 10.30pm    |            |
| Life Changers                          | 4Me        | 11pm       |            |
| Secret to Six Packs!                   | 4Me        | 11.30pm    |            |
| When Love Comes to Town                | Nine       | 8.48pm     |            |
| The Big Bang Theory                    | Nine       | 10.20pm    |            |
| The Big Bang Theory                    | Nine       | 10.48pm    |            |
| Two and a Half Men                     | Nine       | 11.14pm    |            |
| Two and a Half Men                     | Nine       | 11.42pm    |            |
| The Enforcer                           | GEM        | 9.30pm     |            |
| Longmire                               | GEM        | 11.30pm    |            |
| The Town                               | GO!        | 9.30pm     |            |
| Ross Kemp On Gangs                     | ONE        | 9.30pm     |            |
| Super Rugby Extra Time                 | ONE        | 10.30pm    |            |
| The Project                            | ONE        | 11.30pm    |            |
| NCIS: Los Angeles                      | TEN        | 9.30pm     |            |
| TEN Eyewitness News Late               | TEN        | 10.30pm    |            |
| Under the Dome                         | TEN        | 11pm       |            |
| Snog, Marry, Avoid?                    | ELEVEN     | 9.30pm     |            |
| Snog, Marry, Avoid?                    | ELEVEN     | 10.05pm    |            |
| Offspring                              | ELEVEN     | 10.40pm    |            |
| The Late Late Show with Craig Ferguson | ELEVEN     | 11.40pm    |            |
| NEW Destination Gold Launch            | TVSN       | 9.30pm     |            |
| Koalabi Uggs Season Collection         | TVSN       | 10.30pm    |            |
| Cordelia St New Fashion                | TVSN       | 11.30pm    |            |
| Dr Ho's Muscle Therapy                 | SpreeTV    | 10pm       |            |
| Shark 5-in-1 LiftAway Steam Cleaner    | SpreeTV    | 10.30pm    |            |
| Ninja Mega Kitchen System - BL770NZ    | SpreeTV    | 11pm       |            |
| iRobot Roomba                          | SpreeTV    | 11.30pm    |            |
| NEW Destination Gold Launch            | TVSN       | 9.30pm     |            |
| Koalabi Uggs Season Collection         | TVSN       | 10.30pm    |            |
| Cordelia St New Fashion                | TVSN       | 11.30pm    |            |
| Al Jazeera English                     | TVS        | 10pm       |            |
| In Good Shape                          | TVS        | 10.30pm    |            |
| Vasili's Garden to Kitchen             | TVS        | 11pm       |            |
| InStyler - Cool Touch Multi Styler     | FreshIdeas | 9.55pm     |            |
| Derma Wand                             | FreshIdeas | 10.23pm    |            |
| Ninja Mega Kitchen System - BL770NZ    | FreshIdeas | 10.52pm    |            |
| Thin Lizzy 6 in 1 Professional Powder  | FreshIdeas | 11.20pm    |            |
| Pearl Hair Remover                     | FreshIdeas | 11.22pm    |            |
| Aeropilates                            | FreshIdeas | 11.51pm    |            |
| Hot Shapers                            | Extra      | 10pm       |            |
| Jump Start Soup Mate Pro               | Extra      | 10.30pm    |            |
| Bowflex                                | Extra      | 11pm       |            |
| Dr Ho's Pain Therapy System            | Extra      | 11.30pm    |            |
| Dr Ho's Pain Therapy System            | Extra2     | 10pm       |            |
| Ninja Mega Kitchen System              | Extra2     | 10.30pm    |            |
| Total Gym XLS                          | Extra2     | 11pm       |            |
| Focus T25                              | Extra2     | 11.30pm    |            |
+----------------------------------------+------------+------------+------------+
```


changelog
# Changes:
# 0.0.1:
# Initial version
# 0.0.2: Fri 26 Jul 2013 22:25:41 EST
# Updates and initial commit to github
# 0.0.3: Fri 26 Jul 2013 22:52:59 EST
# Added version information
# 0.0.4: Sat 27 Jul 2013 00:29:03 EST
# Added terminal tables
# 0.0.5: Sat 27 Jul 2013 08:39:16 EST
# Used Mechanize to fetch rest of day's programming
# 0.0.6: Sat 27 Jul 2013 10:47:14 EST
# Search based on programming content
# 0.0.7: Sun 28 Jul 2013 21:17:53 EST
# Added handling for empty program information
# 0.0.8: Mon 29 Jul 2013 06:10:10 EST
# Added capability to show staff picks
# 0.0.9: Tue 30 Jul 2013 09:22:21 EST
# Minor updates
# 0.1.0: Sun 13 Oct 2013 19:57:13 EST
# Minor updates
# 0.1.1: Tue 22 Oct 2013 12:11:07 EST
# Improved table title and fixed midday time error
# 0.1.2: Tue 22 Oct 2013 12:47:43 EST
# Added date to programme information
# 0.1.3: Sun 27 Oct 2013 21:21:58 EST
# Fixed option handling to print help when no options given
# 0.1.4: Sat  9 Nov 2013 20:27:30 EST
# Added support for NITV having channel name in programme title
# 0.1.5: Mon 27 Jan 2014 17:58:11 EST
# Fixed bug with am/pm detection
# 0.1.6: Mon 27 Jan 2014 20:00:27 EST
# Improved time handling
# 0.1.7: Tue 20 May 2014 22:15:00 EST
# Added -R switch to show what's on for the rest of the day
# 0.1.8: Tue 17 Feb 2015 19:41:03 AEDT
# Fixed handling of ABC, ABC1, ABC2, and ABC4

wot.rb
#!/usr/bin/env ruby

# Name:         wot (What's On TV)
# Version:      0.1.8
# Release:      1
# License:      Open Source
# Group:        System
# Source:       N/A
# URL:          https://github.com/richardatlateralblast/wot
# Distribution: UNIX
# Vendor:       Lateral Blast
# Packager:     Richard Spindler <richard@lateralblast.com.au>
# Description:  Ruby script to process Australian TV guide

# Load gems

require 'rubygems'
require 'nokogiri'
require 'open-uri'
require 'getopt/std'
require 'terminal-table'
require 'mechanize'

# Set up some defaults

$script        = $0
location       = "melbourne"
time_search    = ""
channel_search = ""
$options       = "c:a:l:s:hnrCLNRV"

def print_usage
  puts ""
  puts "Usage: "+$script+" -["+$options+"]"
  puts ""
  puts "-h: Display usage"
  puts "-c: Channel [TEN,ABC1,..]"
  puts "-a: Time [8,8.00,8:30,..]"
  puts "-l: Location"
  puts "-C: List channels"
  puts "-L: List locations"
  puts "-n: What's on TV now (the current hour)"
  puts "-N: What's on TV next (the next hour)"
  puts "-R: Want's on TV (the rest of the day)"
  puts "-s: Search on subject (eg News)"
  puts "-r: Show only staff picks"
  puts ""
  puts "Example:"
  puts ""
  puts $script+"-c SBS -a 7:30"
  puts ""
  puts "+----------------+----------------+----------------+----------------+"
  puts "|  TV Programme: SBS Melbourne at 7.30 pm (Monday 27 January 2014)  |"
  puts "+----------------+----------------+----------------+----------------+"
  puts "| Program        | Channel        | Time           | Staff Pick     |"
  puts "+----------------+----------------+----------------+----------------+"
  puts "| Strip the City | SBSONE         | 7.30pm         | Yes            |"
  puts "| Strip the City | SBSHD          | 7.30pm         | Yes            |"
  puts "| The Feed       | SBS2           | 7.30pm         |                |"
  puts "+----------------+----------------+----------------+----------------+"
  puts ""
  exit
end

def print_version()
  file_array = IO.readlines $0
  version    = file_array.grep(/^# Version/)[0].split(":")[1].gsub(/^\s+/,'').chomp
  packager   = file_array.grep(/^# Packager/)[0].split(":")[1].gsub(/^\s+/,'').chomp
  name       = file_array.grep(/^# Name/)[0].split(":")[1].gsub(/^\s+/,'').chomp
  puts name+" v. "+version+" "+packager
end

def list_channels(location)
  base_url = "http://www.yourtv.com.au/guide/"
  full_url = base_url+location+"/"
  puts "Channels:"
  puts "--------"
  doc = Nokogiri::HTML(open(base_url))
  doc.css('div.venue-name img').each do |node|
    channel_name = node[:title]
    channel_name = channel_name.gsub(/\s+/,'')
    puts channel_name
  end
  return
end

def list_locations(location)
  base_url = "http://www.yourtv.com.au/guide/"
  full_url = base_url+location+"/"
  puts "Locations:"
  puts "---------"
  doc = Nokogiri::HTML(open(base_url))
  doc.css('li a').each do |node|
    if node.to_s.match(/ TV Guide/)
      location_name = node.text.split(/ /)[0]
      puts location_name
    end
  end
  return
end

def process_entry(table,prog_info,info,channel,time,content_search,staff_pick,staff_search)
  if content_search
    if content_search.match(/[A-z|0-9]/)
      if prog_info.match(/#{content_search}/)
        if staff_search == 1
          if staff_pick == "Yes"
            row = [info,channel,time,staff_pick]
            table.add_row(row)
          end
        else
          row = [info,channel,time,staff_pick]
          table.add_row(row)
        end
      end
    end
  else
    if staff_search == 1
      if staff_pick == "Yes"
        row = [info,channel,time,staff_pick]
        table.add_row(row)
      end
    else
      row = [info,channel,time,staff_pick]
      table.add_row(row)
    end
  end
  return(table)
end

def search_tv_page(channel_search,time_search,content_search,location,staff_search,date)
  rows     = []
  base_url = "http://www.yourtv.com.au/guide/"
  full_url = base_url+location+"/"
  # Split out am/pm from search string so we can cover for 6pm and 6:30pm
  if time_search.match(/[0-9]/) and time_search.match(/am|pm/)
    if time_search.match(/am/)
      meridian = "am"
      time_search = time_search.gsub(/am/,'')
    else
      meridian = "pm"
      time_search = time_search.gsub(/pm/,'')
    end
  end
  # Start with Mechanize so we can follow links and don't need to set cookies
  agent = Mechanize.new
  page  = agent.get(base_url)
  page  = agent.page.link_with(:text => 'Rest of today').click
  page  = page.content
  # Hand over to Nokogiri for processing
  doc = Nokogiri::HTML(page)
  uc_location = location.capitalize
  search_info = "#{channel_search} #{uc_location} #{time_search} #{meridian} #{content_search} (#{date})"
  search_info = search_info.gsub(/\s+/,' ')
  table = Terminal::Table.new :title => "TV Programme: #{search_info}", :headings => ['Program','Channel','Time','Staff Pick']
  doc.css('div.pname').each do |node|
    if node.to_s.match(/Staff Pick/)
      staff_pick = "Yes"
    else
      staff_pick = ""
    end
    prog_info = node.css('a')[0]["title"]
    (info,time) = prog_info.split(/ at /)
    if info.match(/NITV on/)
      (prefix,info,channel) = info.split(/ on /)
      info = prefix+" on "+info
    else
      (info,channel) = info.split(/ on /)
    end
    # Handle Five in the time string for Seven and Nine 5pm news
    # in case you watch this commercial garbage which has half the news in twice the time
    if time.match(/Five/)
      (time,channel) = time.split(/ on /)
      time = "5pm"
    end
    if channel
      channel = channel.gsub(/\s+/,'')
      if channel.match(/\//)
        if channel_search == "ABC2"
          channel = channel.split(/\//)[0]
        else
          channel = channel.split(/\//)[1]
        end          
      end
      if channel_search.match(/[A-z]/)
        if channel == channel_search or channel_search.match(/ALL/)
          if time_search.match(/[0-9]/)
            if time_search.match(/\./)
              if prog_info.match(/#{time_search}#{meridian}/)
                table = process_entry(table,prog_info,info,channel,time,content_search,staff_pick,staff_search)
              end
            else
              if prog_info.match(/#{time_search}\.[0-9][0-9]#{meridian}/) or prog_info.match(/#{time_search}#{meridian}/)
                table = process_entry(table,prog_info,info,channel,time,content_search,staff_pick,staff_search)
              end
            end
          else
            table = process_entry(table,prog_info,info,channel,time,content_search,staff_pick,staff_search)
          end
        end
      else
        if time_search.match(/[0-9]/)
          if time_search.match(/\./)
            if prog_info.match(/#{time_search}#{meridian}/)
              table = process_entry(table,prog_info,info,channel,time,content_search,staff_pick,staff_search)
            end
          else
            if prog_info.match(/#{time_search}\.[0-9][0-9]#{meridian}/) or prog_info.match(/#{time_search}#{meridian}/)
              table = process_entry(table,prog_info,info,channel,time,content_search,staff_pick,staff_search)
            end
          end
        else
          table = process_entry(table,prog_info,info,channel,time,content_search,staff_pick,staff_search)
        end
      end
    end
  end
  puts table
  return
end

def handle_channel(channel_search)
  channel_search = channel_search.gsub(/\s+/,'')
  case channel_search
  when /^7$/
    channel_search = "Seven"
  when /^9$/
    channel_search = "Nine"
  when /^2$/
    channel_search = "ABC2"
  when /^1$|ABC1|ABC 1|^ABC$/
    channel_search = "ABC"
  when /SBS1|SBS 1/
    channel_search = "SBSONE"
  when /^0$/
    channel_search = "SBSONE"
  when /SBS2|SBS 2/
    channel_search = "SBS2"
  when /^4$/
    channel_search = "ABC4"
  when /10$/
    channel_search = "TEN"
  when /11$/
    channel_search = "ELEVEN"
  when /24$/
    channel_search = "ABCNews24"
  end
  if !channel_search.downcase.match(/news/)
    channel_search = channel_search.downcase
    if !channel_search.match(/mate/)
      if channel_search.match(/seven|nine/)
        channel_search = channel_search.capitalize
      else
        channel_search = channel_search.upcase
      end
    end
  else
    channel_search = "ABCNews24"
  end
  return(channel_search)
end

# Get command line arguments
# Print help if given none

if !ARGV[0]
  print_usage()
end

begin
  opt = Getopt::Std.getopts($options)
rescue
  print_usage()
end

if opt["V"]
  print_version
  exit
end

# Handle getting what is on in the current or next hour

if opt["n"] or opt["N"]
  time = Time.new
  hour = time.hour
  if hour > 11
    hour = hour-12
    suffix = "pm"
  else
    suffix = "am"
  end
  if hour == 0
    hour = 12
    suffix = "am"
    if opt["N"]
      hour = 1
      suffix = "pm"
    end
  else
    if opt ["N"]
      hour = hour+1
    end
  end
  opt["a"] = hour.to_s+suffix
end

# Ouput loations (cities)

if opt["L"]
  list_locations(location)
  exit
end

# output channels

if opt["C"]
  list_channels(location)
  exit
end

# Handle location

if opt['l']
  location = opt["l"].downcase
end

# Handle channel

if opt["c"]
  channel_search = opt["c"]
  channel_search = handle_channel(channel_search)
end

# Handle time search

def handle_time_search(time_search)
  time_search = time_search.sub(":",".")
  if !time_search.downcase.match(/am|pm/)
    if time_search.match(/\./)
      hour = time_search.split(/\./)[0]
      mins = time_search.split(/\./)[1]
    else
      hour = time_search
      mins = "00"
    end
    if hour.to_i > 12
      hour = hour.to_i - 12
      hour = hour.to_s
      meridian = "pm"
    else
      if hour.to_i == 0
        hour = "12"
        meridian = "am"
      else
        meridian = Time.now.strftime('%p').downcase
      end
    end
    if !mins.match(/00/)
      time_search = hour+"."+mins
    else
      time_search = hour
    end
    time_search = time_search+meridian
  end
  time_search = "at "+time_search
  return time_search
end

# Handle time

if opt["a"]
  time_search = opt["a"]
  time_search = handle_time_search(time_search)
end

# Staff picks only

if opt["r"]
  staff_search = 1
else
  staff_search = 0
end

# Handle content

if opt["s"]
  content_search = opt["s"]
end

if opt["a"] or opt["c"] or opt["s"] or opt["r"] or opt["R"]
  time = Time.new
  date = time.strftime("%A %d %B %Y")
  if opt["R"]
    time = ""
  end
  search_tv_page(channel_search,time_search,content_search,location,staff_search,date)
end

if opt["h"]
  print_usage()
end
